neo/tools/imgui/pdaeditor/PDAEditor.cpp

```diff
@@ -0,0 +1,720 @@
+	if ( file ) {
+		fileSystem->CloseFile( file );
+		saveButtonEnabled = true;
+	} else {
+		windowTitle += " [Read Only]";
+		saveButtonEnabled = false;
+	}
+
+	num = pda->GetNumEmails();
+	emailList.Resize( num );
+	emailList.Clear();
+	emailListSel = -1;
+	for ( i=0; i < num; i++ ) {
+		emailList.Append( pda->GetEmailByIndex( i )->GetSubject() );
+	}
+
+	num = pda->GetNumAudios();
+	audioList.Resize( num );
+	audioList.Clear();
+	audioListSel = -1;
+	for ( i=0; i < num; i++ ) {
+		audioList.Append( pda->GetAudioByIndex( i )->GetAudioName() );
+	}
+
+	num = pda->GetNumVideos();
+	videoList.Resize( num );
+	videoList.Clear();
+	videoListSel = -1;
+	for ( i=0; i < num; i++ ) {
+		videoList.Append( pda->GetVideoByIndex( i )->GetVideoName() );
+	}
+
+	fullName = pda->GetFullName();
+	shortName = pda->GetPdaName();
+	post = pda->GetPost();
+	title = pda->GetTitle();
+	security = pda->GetSecurity();
+	idnum = pda->GetID();
+}
+
+void PDAEditor::OnBtnClickedSave()
+{
+	int index = pdaListSel;
+	if ( index < 0 ) {
+		return;
+	}
+
+	const idDeclPDA *pdaConst = dynamic_cast<const idDeclPDA *>( declManager->DeclByIndex(DECL_PDA, index) );
+	if ( pdaConst ) {
+		idDeclPDA *pda = const_cast<idDeclPDA *>(pdaConst);
+
+		idStr declText = "\n";
+		declText += "pda ";
+		declText += pda->GetName();
+		declText += " {\n";
+
+		declText += "\tname    \t\t\"" + shortName + "\"\n";
+		declText += "\tfullname\t\t\"" + fullName + "\"\n";
+		declText += "\ticon    \t\t\"\"\n";
+		declText += "\tid      \t\t\"" + idnum + "\"\n";
+		declText += "\tpost    \t\t\"" + post + "\"\n";
+		declText += "\ttitle   \t\t\"" + title + "\"\n";
+		declText += "\tsecurity\t\t\"" + security + "\"\n";
+
+		for ( int i = 0; i < pda->GetNumEmails(); i++ ) {
+			declText += "\tpda_email\t\t\"";
+			declText += pda->GetEmailByIndex(i)->GetName();
+			declText += "\"\n";
+		}
+
+		for ( int i = 0; i < pda->GetNumAudios(); i++ ) {
+			declText += "\tpda_audio\t\t\"";
+			declText += pda->GetAudioByIndex(i)->GetName();
+			declText += "\"\n";
+		}
+
+		for ( int i = 0; i < pda->GetNumVideos(); i++ ) {
+			declText += "\tpda_video\t\t\"";
+			declText += pda->GetVideoByIndex(i)->GetName();
+			declText += "\"\n";
+		}
+
+		declText += "}";
+
+		pda->SetText( declText );
+		pda->ReplaceSourceFileText();
+		pda->Invalidate();
+	}
+}
+
+DialogPDAAdd::DialogPDAAdd() {
+	Reset();
+}
+
+void DialogPDAAdd::Reset() {
+	name.Clear();
+}
+
+bool DialogPDAAdd::Draw() {
+	ImGui::InputTextStr( "Name", &name );
+
+	bool isAccepted = false;
+	if ( ImGui::Button( "OK" ) ) {
+		isAccepted = true;
+	}
+	ImGui::SameLine();
+	if ( ImGui::Button( "Cancel" ) ) {
+		isAccepted = false;
+	}
+	return isAccepted;
+}
+
+
+void PDAEditor::OnBtnClickedRandom()
+{
+	idnum = idStr::Format("%d-%02X", 1000+(rand()%8999), (rand()%255));
+}
+
+void PDAEditor::OnBtnClickedPDAAdd()
+{
+	addPDADlg.Reset();
+	ImGui::OpenPopup( "PDAAdd" );
+}
+
+void PDAEditor::OnBtnClickedPDADel()
+{
```

Comment:
maybe this could be implemented like
```c++
idStr fileName = "newpdas/";
fileName += pdaName; // TODO
fileName += ".pda";
fileSystem->RemoveFile( fileName );
```